### ISSUE TITLE
TEAL - Review program size limits with real programs

### ISSUE THREAD

**Issue by jeapostrophe** [algorand/go-algorand#1005]
I'm looking at the limits here:

https://github.com/justicz/go-algorand/blob/a3200f5e0b6485629191ad844612e136336193bf/config/consensus.go#L636

And I think it is very important for these to be evaluated against real target programs before they are finalized.

For example, I have a really basic program that I'm compiling from our language, Reach, that is at 1587 bytes for the approval program, uses 11 arguments, 1 global state variable, and each method takes about 138 cost.

Given this is the most basic program in our language, I'm very worried about these limits. In particular, we're already over the size limit (which is 1024) and we're really close to the argument limit (11 out of 16), and we're at 20% of the cost limit.

In general, I think these limits should be chosen to be as big as possible for the constraints of the actual network, rather than arbitrarily chosen. I want to see a graph or a system of linear equations that says, "Because of the speed of light and the consensus timeout, we can only have X bytes for the program size", etc. 

Similarly, I want to have as few limits as possible. For example, right now there is a limit to the number of arguments and their length, which are independent limits. This means that if I get to a situation where I need more arguments, I may have to cram multiple arguments into one; or symmetrically, if I need more length, I may have to split one argument in to many. Instead, if there were only a limit on the total length of all the arguments, then I would not need to play these games.

Are `MaxAppProgramCost` and `MaxApprovalProgramLen` similar? Is it necessary to put a limit on the size and the cost independently? My guess is that the size limit comes from the expense of replicating the data across the network on a creation event. It seems hard to believe that that is 1024. My guess is that the cost limit comes from the consensus time; again 700 seems like it is not "real" because it is really cheap on even the most basic hardware to do 700 64-bit adds (a cost 1 operation).

In summary, I really want to be able to deploy on Algorand and so I would like the limits to be based on real applications, like those that I'm compiling and writing.

Thank you!

**Comment by derbear** [comment]
We really appreciate your feedback--we're still rapidly iterating and actively trying to raise these limits on the prototype branch you've pointed to. It would be great if you could post a source program you're trying to compile and the corresponding TEAL assembly compiler output! A concrete example would be extremely helpful input for us in choosing our limits.

**Comment by jeapostrophe** [comment]
The program I'm talking about is [here](https://github.com/reach-sh/reach-lang/blob/master/examples/rps/rsh/rps.rsh) and the compiled TEAL code is [here](https://github.com/reach-sh/reach-lang/blob/master/examples/rps/build/rps.teal).

However, I have not yet been able to build the Algorand standard library for our language (since there's no SDKs for the application prototype), so I have not tested it. I had a meeting with @algobolson yesterday and learned that there are a few problems with my TEAL code (e.g., I use `arg N` rather than `txna ApplicationArgs N`, because I didn't realized logic signature arguments and application arguments are distinct.)

I'd be happy to get on the phone/etc and talk about this problem with anyone on your team :)

**Comment by jannotti** [comment]
The first approach is probably to stop enforcing many separate individual limits, when the real goal is some higher level limit.  Two specifics: LogicSigs need some limit on their size, mostly in pursuit of the higher level limit of keeping total transaction group size limited, to simplify block proposal.  So we should move toward enforcing limits on total transaction (group) size rather than individual limits.  Since we care about the size *after* execution of any TEAL apps (which add to ApplyData), this is a little trickier than a simple limit on submission size, but not much.

For apps, the higher level limit we want to enforce is on the amount of storage the blockchain provides. So, again, there's no reason for enforcement of separate limits on approval and clear state program sizes. And we should also consider allowing larger programs in return for larger minimum balance increments, just as we do for schema state.

**Comment by algoanne** [comment]
I love this old issue. I'm going to close it as it's too general to be useful at this point - we 100% agree with the sentiment and continue to progressively strive for a perfectly balanced system where the limits are _just_ as limiting as they need to be.
